homeassistant/components/fronius/config_flow.py

```diff
@@ -0,0 +1,116 @@
+"""Config flow for Fronius integration."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from pyfronius import Fronius, FroniusError
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_HOST, CONF_RESOURCE
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import DOMAIN, FroniusConfigEntryData
+
+_LOGGER = logging.getLogger(__name__)
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): str,
+    }
+)
+
+
+async def validate_input(
+    hass: HomeAssistant, data: dict[str, Any]
+) -> tuple[str, FroniusConfigEntryData]:
+    """Validate the user input allows us to connect.
+
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+    """
+    host = data[CONF_HOST]
+    fronius = Fronius(async_get_clientsession(hass), host)
+
+    try:
+        logger_info: dict[str, Any]
+        logger_info = await fronius.current_logger_info()
+    except FroniusError as err:
+        _LOGGER.debug(err)
+    else:
+        logger_uid: str = logger_info["unique_identifier"]["value"]
+        return logger_uid, FroniusConfigEntryData(
+            host=host,
+            is_logger=True,
+        )
+    # Gen24 devices don't provide GetLoggerInfo
+    try:
+        inverter_info = await fronius.inverter_info()
+        first_inverter = next(inverter for inverter in inverter_info["inverters"])
+    except FroniusError as err:
+        _LOGGER.debug(err)
+        raise CannotConnect from err
+    except StopIteration as err:
+        raise CannotConnect("No supported Fronius SolarNet device found.") from err
+    first_inverter_uid: str = first_inverter["unique_id"]["value"]
+    return first_inverter_uid, FroniusConfigEntryData(
+        host=host,
+        is_logger=False,
+    )
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Fronius."""
+
+    VERSION = 1
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=STEP_USER_DATA_SCHEMA
+            )
+
+        errors = {}
+
+        try:
+            unique_id, info = await validate_input(self.hass, user_input)
+        except CannotConnect:
+            errors["base"] = "cannot_connect"
+        except Exception:  # pylint: disable=broad-except
```

Comment:
When will this be thrown? You're handling all the cases already in `validate_input` or am I missing something?

Reply:
only `FroniusError` is chatched in `validate_input`. This catches unexpected errors - such as some KeyError happening in pyfronius (which hopefully is handled there anyway, but you never know 👮‍♂️)

Reply:
Alright :)